src/task.rs

```diff
@@ -38,6 +37,56 @@ impl PartialOrd for Task {
     }
 }
 
+//An iterator for start/deadline combinations for the task.
+//e.g. if a task has duration of 2, and one slot with start 10 and end 14,
+//then the following start/deadline combinations are possible:
+//10-12, 11-13, and 12-14.
+//It also needs to handle scenarios where the task has multiple slots.
+pub struct StartDeadlineIterator {
+    slots: Vec<Slot>,
+    duration: usize,
+    marker: NaiveDateTime,
+}
+
+impl StartDeadlineIterator {
+    fn new(slots: Vec<Slot>, duration: usize) -> StartDeadlineIterator {
+        let marker = slots[0].start;
+        StartDeadlineIterator {
+            slots,
+            duration,
+            marker,
+        }
+    }
+}
+
+impl Iterator for StartDeadlineIterator {
+    type Item = Slot;
+    fn next(&mut self) -> Option<Self::Item> {
+        for (index, slot) in self.slots.iter().enumerate() {
+            if !(self.marker >= slot.start && self.marker < slot.end) {
+                continue;
+            }
+            while (self.marker + Duration::hours(self.duration as i64)) <= slot.end {
+                let start = self.marker;
+                let end = self.marker + Duration::hours(self.duration as i64);
+                self.marker += Duration::hours(1);
+                if self.marker == slot.end {
+                    if index != self.slots.len() - 1 {
+                        self.marker = self.slots[index + 1].start;
+                    }
+                }
+                return Some(Slot { start, end });
+            }
+
+            if index != self.slots.len() - 1 {
+                self.marker = self.slots[index + 1].start;
+            }
+        }
+
+        return None;
+    }
+}
```

Comment:
```suggestion
pub struct StartDeadlineIterator {
    slots: Vec<Slot>,
    duration: usize,
    marker: NaiveDateTime,
    slot_index: usize,
}

impl StartDeadlineIterator {
    fn new(slots: Vec<Slot>, duration: usize) -> StartDeadlineIterator {
        let marker = slots[0].start;
        let slot_index = 0 as usize;
        StartDeadlineIterator {
            slots,
            duration,
            marker,
            slot_index,
        }
    }
}

impl Iterator for StartDeadlineIterator {
    type Item = Slot;
    fn next(&mut self) -> Option<Self::Item> {
        if self.marker > self.slots[self.slot_index].end - Duration::hours(self.duration as i64) {
            if self.slot_index != self.slots.len() - 1 {
                self.marker = self.slots[self.slot_index + 1].start;
                self.slot_index += 1;
            } else {
                return None;
            }
        }
        let start = self.marker;
        let end = self.marker + Duration::hours(self.duration as i64);
        self.marker += Duration::hours(1);
        return Some(Slot { start, end });
    }
}
```

I found the iteration a bit hard to read.
Whenever I see ```if !(something)``` I like to refactor to the positive form.
That didn't solve the readability.

I suggest we not only keep a marker to the current NaiveDateTime but also a marker to the slot.
This way we don't have to loop through all the slots at every ```next``` and it's more readable?